app/exec/build/tasks/sign.ts

```diff
@@ -0,0 +1,159 @@
+import admZip = require("adm-zip");
+import archiver = require("archiver");
+import del = require('del');
+import fs = require("fs");
+var ncp = require('ncp').ncp;
+import os = require('os');
+import path = require("path");
+import shell = require("shelljs");
+import tasksBase = require("./default");
+import trace = require("../../../lib/trace");
+
+export interface TaskSignResult {
+	signingSuccessful: boolean;
+}
+
+export function getCommand(args: string[]): BuildTaskSign {
+	return new BuildTaskSign(args);
+}
+
+export class BuildTaskSign extends tasksBase.BuildTaskBase<TaskSignResult> {
+	protected description = "Sign a task.";
+
+	constructor(args: string[]) {
+		super(args);
+	}
+
+	protected getHelpArgs(): string[] {
+		return ["certFingerprint", "taskPath", "newGuid", "newNameSuffix"];
+	}
+
+	// tfx build tasks sign --task-path {TASK_PATH} --cert-fingerprint {FINGERPRINT} --new-guid {GUID} --new-name-suffix ' - SIGNED'
+	// --new-guid and --new-name-suffix are optional
+	public async exec(): Promise<TaskSignResult> {
+		const taskZipsPath: string[] | null = await this.commandArgs.taskPath.val();
+		const taskZipPath = taskZipsPath[0];
+		const certFingerprint: string | null = await this.commandArgs.certFingerprint.val();
+		const newGuid: string | null = await this.commandArgs.newGuid.val();
+		const newNameSuffix: string | null = await this.commandArgs.newNameSuffix.val();
+
+		// verify that we can find NuGet
+		const nuGetPath: string = shell.which('nuget');
+		if (!nuGetPath) {
+			throw new Error('Unable to find NuGet. Please add NuGet to the PATH before continuing.');
+		}
+
+		const tempFolder: string = fs.mkdtempSync(path.join(os.tmpdir(), 'task-'));
+		let taskTempFolder: string = path.join(tempFolder, 'task');
+		const taskTempZipPath: string = path.join(tempFolder, 'task.zip');
+		const taskTempNupkgPath: string = path.join(tempFolder, 'task.nupkg');
+
+		// Create temp folder
+		if (fs.existsSync(tempFolder)) {
+			await del(tempFolder, { force: true });
+		}
+
+		fs.mkdirSync(tempFolder);
+		fs.mkdirSync(taskTempFolder);
+
+		// Copy task contents to temp folder
+		shell.cp('-R', taskZipPath, taskTempFolder);
```

Comment:
I think we can do without this copy if we just create a unzip straight from `taskZipPath` to the temp folder

Reply:
I don't thin it overrides files when you extract directly. THis gives us more control.